feat(core): add controllers, services, manual mappers, global error handler and tests

a-buianova opened the pull request:
### What’s inside
- Users/Items controllers (CRUD + search)
- Manual DTO mappers (no MapStruct)
- In-memory repositories (users/items)
- GlobalExceptionHandler with unified error JSON
- Custom exceptions (400/403/404/409)
- Header resolver @CurrentUserId
- Unit + web layer tests (MockMvc), context smoke test

### Why
- Keep build simple and fast for sprint 14
- Prepare groundwork for bookings in next sprint

### How to test
```bash
mvn verify

a-buianova committed:
feat: implement core controllers, services, mappers, error handling and tests

Ivanov-s-tmn commented on src/main/java/ru/practicum/shareit/AppConfig.java:
```diff
@@ -0,0 +1,23 @@
+/**
+ * @apiNote Central application configuration for in-memory repositories (sprint 14 only).
+ */
+@Configuration
+public class AppConfig {
+
+    @Bean
+    public InMemoryUserRepository userRepository() {
+        return new InMemoryUserRepository();
+    }
+
+    @Bean
+    public InMemoryItemRepository itemRepository() {
+        return new InMemoryItemRepository();
+    }
+}
```
А зачем описывать бин методы для компонентов? Они создадутся сами и внедрятся спрингом

a-buianova replied:
Да, согласна. Изначально вынесла @Bean для явного контроля и чтобы было видно, какие компоненты участвуют в контексте. Но т.к. классы уже аннотированы как @Repository, это действительно дублирование. Уберу лишние бин-методы.

Ivanov-s-tmn commented on src/main/java/ru/practicum/shareit/booking/mapper/BookingMapper.java:
```diff
@@ -0,0 +1,47 @@
+                b.getStatus() != null ? b.getStatus().name() : null
+        );
+    }
+}
```
Нет особого смысла делать маппер компонентом, можно просто сделать его методы статическими, вызывать где угодно и при этом не перегружать контекст. Если бы мапперы были реализованы через MapStruct, то без компонента уже не обойтись

Ivanov-s-tmn commented on src/main/java/ru/practicum/shareit/item/controller/ItemController.java:
```diff
@@ -0,0 +1,68 @@
+    @PatchMapping("/{itemId}")
+    public ItemResponse patch(@CurrentUserId Long userId,
+                              @PathVariable Long itemId,
+                              @RequestBody @Valid ItemUpdateDto dto) {
```
В `ItemUpdateDto` нет валидации

a-buianova replied:
Верно, т.к. это DTO для частичных обновлений, сделала все поля nullable. Добавлю хотя бы базовые аннотации.

Ivanov-s-tmn commented on src/main/java/ru/practicum/shareit/item/repo/InMemoryItemRepository.java:
```diff
@@ -0,0 +1,101 @@
+    public Item save(Item i) {
+        long id = seq.incrementAndGet();
+        i.setId(id);
+        store.put(id, i);
```
Метод изменяет переданный объект, что нарушает принцип иммутабельности и может привести к неожиданному поведению

Лучше работать с копией

Ivanov-s-tmn commented on src/main/java/ru/practicum/shareit/item/repo/InMemoryItemRepository.java:
```diff
@@ -0,0 +1,101 @@
+
+    /** @return snapshot list of owner's items (never null). */
+    public List<Item> findByOwner(Long ownerId) {
+        return new ArrayList<>(byOwner.getOrDefault(ownerId, new CopyOnWriteArrayList<>()));
```
`new CopyOnWriteArrayList<>()` создаётся каждый раз, когда владельца нет
Это бесполезная аллокация

a-buianova replied:
Точно. Спасибо. Лучше возвращать Collections.emptyList() или просто новый ArrayList<>(). Исправлю на более лёгкий вариант.

Ivanov-s-tmn commented on src/main/java/ru/practicum/shareit/item/service/ItemServiceImpl.java:
```diff
@@ -0,0 +1,63 @@
+            throw new ForbiddenException("forbidden: not an owner");
+        }
+        mapper.patch(existing, dto);
+        Item updated = repo.update(existing).orElseThrow(() -> new NotFoundException("item not found"));
```
Очень странно, что update() возвращает Optional

a-buianova replied:
Хотела унифицировать сигнатуры с findById, но для update это действительно избыточно. Переделаю

Ivanov-s-tmn commented on src/main/java/ru/practicum/shareit/item/repo/InMemoryItemRepository.java:
```diff
@@ -0,0 +1,101 @@
+    public List<Item> searchAvailable(String text) {
+        if (text == null) {
+            return List.of();
+        }
```
Эту проверку стоит вынести на уровень выше, в сервис или в контроллер

Ivanov-s-tmn commented on src/test/java/ru/practicum/shareit/booking/BookingControllerMockMvcTest.java:
```diff
@@ -0,0 +1,43 @@
+    @Autowired
+    private MockMvc mvc;
+
+    @Test
+    @DisplayName("GET /bookings -> 405 Method Not Allowed (no endpoints)")
+    void get_onBasePath_returns405() throws Exception {
+        mvc.perform(get("/bookings"))
+                .andExpect(status().isMethodNotAllowed());
+    }
+
+    @Test
+    @DisplayName("POST /bookings -> 405 Method Not Allowed (no endpoints)")
+    void post_onBasePath_returns405() throws Exception {
+        mvc.perform(post("/bookings"))
+                .andExpect(status().isMethodNotAllowed());
+    }
```
Одинаковые тесты?

a-buianova replied:
Да, оба проверяют 405, но для разных HTTP-методов (GET и POST). Хотела явно зафиксировать поведение на оба случая, но понимаю, что это воспринимается как дублирование. Могу либо оставить один тест, либо добавить комментарий о намеренном тестировании разных методов.

a-buianova committed:
refactor: apply review feedback for item, booking and user modules

Ivanov-s-tmn commented on src/main/java/ru/practicum/shareit/item/repo/InMemoryItemRepository.java:
```diff
@@ -18,45 +15,58 @@
 
-    /** @return persisted entity with generated id and owner index updated. */
-    public Item save(Item i) {
+    private static final List<Item> EMPTY = Collections.emptyList();
```
Лучше тогда вынести на локальный уровень метода и просто по факту возвращать `Collections.emptyList()`

Ivanov-s-tmn commented on src/main/java/ru/practicum/shareit/item/repo/InMemoryItemRepository.java:
```diff
@@ -66,36 +76,72 @@ public Optional<Item> update(Item i) {
+                .available(s.isAvailable())
+                .owner(s.getOwner())
+                .request(s.getRequest())
+                .build();
```
Можно объединить добавив простую проверку айди на null и если нужно передать айди, то передаем, если нет - передаем вместо него null

a-buianova committed:
refactor(item): simplify repository internals and cleanup formatting

- InMemoryItemRepository:
  * removed static EMPTY constant, use Collections.emptyList() directly
  * merged copy() and copyWithId() into a single method with idOverride
  * improved formatting, consistent braces and spacing
  * streamlined findByOwner() for clarity
- Unified Javadoc style across methods (short, focused descriptions)
- Minor cosmetic cleanup for readability and consistency
